bundles/org.eclipse.jface/src/org/eclipse/jface/resource/URLImageDescriptor.java

```diff
@@ -42,7 +43,78 @@
  * public API. Use ImageDescriptor#createFromURL to create a descriptor that
  * uses a URL.
  */
-class URLImageDescriptor extends ImageDescriptor implements IAdaptable, ImageFileNameProvider, ImageDataProvider {
+class URLImageDescriptor extends ImageDescriptor implements IAdaptable {
+
+	private static class URLImageFileNameProvider implements ImageFileNameProvider {
+
+		private final String url;
+
+		public URLImageFileNameProvider(String url) {
+			this.url = url;
+		}
+
+		@Override
+		public String getImagePath(int zoom) {
+			URL tempURL = getURL(url);
+			if (tempURL != null) {
+				final boolean logIOException = zoom == 100;
+				if (zoom == 100) {
+					return getFilePath(tempURL, logIOException);
+				}
+				URL xUrl = getxURL(tempURL, zoom);
+				if (xUrl != null) {
+					String xResult = getFilePath(xUrl, logIOException);
+					if (xResult != null) {
+						return xResult;
+					}
+				}
+				String xpath = FileImageDescriptor.getxPath(url, zoom);
+				if (xpath != null) {
+					URL xPathUrl = getURL(xpath);
+					if (xPathUrl != null) {
+						return getFilePath(xPathUrl, logIOException);
+					}
+				}
+			}
+			return null;
+		}
+
+	}
+
+	private static class URLImageDataProvider implements ImageDataProvider {
```

Comment:
What is the reason to duplicate the method implementation here?
Why can't we do something like this:

```
	private class URLImageDataProvider implements ImageDataProvider {
		@Override
		public ImageData getImageData(int zoom) {
			return URLImageDescriptor.this.getImageData(zoom);
		}
	}
```

or extract the main part of `getImageData(zoom)` to a static method and do something like this:

```
	private static class URLImageDataProvider implements ImageDataProvider {
		private final String url;

		public URLImageDataProvider(String url) {
			this.url = url;
		}
		
		@Override
		public ImageData getImageData(int zoom) {
			return URLImageDescriptor.getImageData(zoom, url);
		}

	}

	@Override
	public ImageData getImageData(int zoom) {
		return getImageData(zoom, url);
	}
	
	static ImageData getImageData(int zoom, String url) {
		URL tempURL = getURL(url);
		if (tempURL != null) {
			if (zoom == 100) {
				return getImageData(tempURL);
			}
			URL xUrl = getxURL(tempURL, zoom);
			if (xUrl != null) {
				ImageData xdata = getImageData(xUrl);
				if (xdata != null) {
					return xdata;
				}
			}
			String xpath = FileImageDescriptor.getxPath(url, zoom);
			if (xpath != null) {
				URL xPathUrl = getURL(xpath);
				if (xPathUrl != null) {
					return getImageData(xPathUrl);
				}
			}
		}
		return null;
	}
```